### PR TITLE
Deprecate auto-removal of overlapping Axes by plt.subplot{,2grid}.

### LINKED ISSUE #4976: missing imshow() subplots when using tight_layout()
Hello, I've found an unexpected behavior when I make grids of subplots with imshow and tight_layout the second to last plot is missing, for example:

```
import numpy as np
import matplotlib.pyplot as plt
plt.figure()
for i in range(16):
    ax= plt.subplot(4,4 ,i+1)
    im=ax.imshow(np.random.normal(size=100).reshape([10,10]))
    plt.tight_layout()
    plt.title(i)
plt.savefig("imshow_4x4.png")
```

generates:
imshow_4x4
Whereas the same loop with pcolor works:

```
import numpy as np
import matplotlib.pyplot as plt
plt.figure()
for i in range(16):
    ax= plt.subplot(4,4 ,i+1)
    im=ax.pcolor(np.random.normal(size=100).reshape([10,10]))
    plt.tight_layout()
    plt.title(i)
plt.savefig("pcolor_4x4.png")
```

gives:
pcolor_4x4

If I omit the tight_layout() call from the first example it works:
imshow_nottight_4x4

I'm using matplotlib 1.4.3 and numpy  1.9.2  in Anaconda 2.2.0.

### LINKED ISSUE #11435: plt.subplot eats my subplots
### Bug report

**Bug summary**

`plt.subplot` eats my subplots. But it doesn't completely digest them, since they are still there, just not shown.

Creating a subplot grid with `plt.subplots` and in addition adding a `plt.subplot` removes those subplots from the grid, which are under the newly created subplot.  
Creating the additional subplot via `fig.add_subplot` does not remove them.

**Code for reproduction**

```
import matplotlib.pyplot as plt
fig, ax = plt.subplots(2,3,
                       subplot_kw={"label" : "A"}, 
                       gridspec_kw={'width_ratios':[1,1,3]})
ax[0,2].plot([1,2,3])

axi = plt.subplot(133, label="B")
axi.plot([7,6,5], color="red")
axi.axis("off")

plt.show()
```

**Actual outcome**

image

**Expected outcome**

The expected outcome would be the same as with `fig.add_subplot`:

```
axi = fig.add_subplot(133, label="B")
axi.plot([7,6,5], color="red")
axi.axis("off")
```

image

**Matplotlib version**

  * Operating system: Win 8.1
  * Matplotlib version: master
  * Matplotlib backend: Qt5Agg
  * Python version: 3.6

### PR DESCRIPTION
**Pull request by anntzer**
In particular, note that the OO add_subplot does not have this behavior.

Closes #4976.  Closes #11435 (which was deemed "not a bug", but still...)

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Review by timhoffm — COMMENTED** [review]
If something is intentionally built in, there's usually a reason. Do have an idea why people might want to replace fully overlapped axes? Is this maybe MATLAB behavior?

I agree that this behavior is quite unexpected, and I'm +0.8 on removing. I just want to make sure we're not busting a valid use case. While you can remove fully overlapping axes manually, that's quite a bit of an effort, in particular if you come from the "simple" pyplot interface. The alternative to a full deprecation would be to make this configurable via a parameter and migrate the default to not removing.

**Comment by anntzer** [comment]
I don't think it comes from MATLAB (https://github.com/matplotlib/matplotlib/pull/7377#issuecomment-260049527)

**Comment by anntzer** [comment]
Ah, good catch, I just check that it does...

**Comment by anntzer** [comment]
I don't think(?) we deprecated that before.

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/tests/test_pyplot.py [hunk 1]
```diff
@@ -206,7 +206,7 @@ def test_subplot_replace_projection():
     ax = plt.subplot(1, 2, 1)
     ax1 = plt.subplot(1, 2, 1)
     ax2 = plt.subplot(1, 2, 2)
-    # This will delete ax / ax1 as they fully overlap
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_pyplot.py:209]: At least one should be kept and the warning captured.

Comment (anntzer) [reply lib/matplotlib/tests/test_pyplot.py:209]: sure

**Comment by QuLogic** [comment]
It appears tests don't like it?

**Comment by anntzer** [comment]
oops, should be fixed.

**Review by jklymak — APPROVED** [review]

**Pull request merged**